From 492aa2b6a91eb32b154b7ab84c1ee581720a09ad Mon Sep 17 00:00:00 2001
From: Camelia Groza <camelia.groza@nxp.com>
Date: Wed, 8 Nov 2017 14:37:27 +0200
Subject: [PATCH 175/741] sdk_dpaa: ls1043a errata: stop advertising S/G and
 GSO support

commit cee2ac9e26ec07c37f2acfb560a19fd402d95d3f from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

The errata prevents us from transmitting S/G frames. Instead of
linearizing them ourselves, stop advertising S/G support and have the
stack do it for us.

Signed-off-by: Camelia Groza <camelia.groza@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 .../ethernet/freescale/sdk_dpaa/dpaa_eth.c    | 21 +++++++++++--------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth.c b/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth.c
index e39e2f009ffa..296510911687 100644
--- a/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth.c
+++ b/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth.c
@@ -761,15 +761,6 @@ static int dpa_private_netdev_init(struct net_device *net_dev)
 	net_dev->min_mtu = ETH_MIN_MTU;
 	net_dev->max_mtu = dpa_get_max_mtu();
 
-#ifndef CONFIG_PPC
-	/* Due to the A010022 FMan errata, we can not use contig frames larger
-	 * than 4K, nor S/G frames. We need to prevent the user from setting a
-	 * large MTU.
-	 */
-	if (unlikely(dpaa_errata_a010022))
-		net_dev->max_mtu = DPA_BP_RAW_SIZE;
-#endif
-
 	net_dev->hw_features |= (NETIF_F_IP_CSUM | NETIF_F_IPV6_CSUM |
 		NETIF_F_LLTX);
 
@@ -787,6 +778,18 @@ static int dpa_private_netdev_init(struct net_device *net_dev)
 	/* Advertise NETIF_F_HW_ACCEL_MQ to avoid Tx timeout warnings */
 	net_dev->features |= NETIF_F_HW_ACCEL_MQ;
 
+#ifndef CONFIG_PPC
+	/* Due to the A010022 FMan errata, we can not use contig frames larger
+	 * than 4K, nor S/G frames. We need to prevent the user from setting a
+	 * large MTU. We also stop advertising S/G and GSO support.
+	 */
+	if (unlikely(dpaa_errata_a010022)) {
+		net_dev->max_mtu = DPA_BP_RAW_SIZE;
+		net_dev->hw_features &= ~NETIF_F_SG;
+		net_dev->features &= ~NETIF_F_GSO;
+	}
+#endif
+
 	return dpa_netdev_init(net_dev, mac_addr, tx_timeout);
 }
 
-- 
2.17.1

